### PR TITLE
plugin-storage-volume-linstor: support QoS(IOPs) and small improvements

### PR DESCRIPTION
**Pull request by rp-**
### Description

This PR has 3 improvements for the Linstor primary storage driver:

- Create a separate jar of it and move all Linstor related classes into the correct project (similar to the storpool plugin)
- Add aux properties for Cloudstack volumes in Linstor to make it easier to identify them in Linstor
- Add support for IOPs settings with the Linstor storage plugin

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
Local test cluster with the Linstor primary storage driver in place.
Creating VMs and volumes and moving them between different disk-offerings.

### CODE REVIEW

**Comment by rohityadavcloud** [comment]
@rp- can you check the security hotspot?

**Comment by rp-** [comment]
@rohityadavcloud btw. I fixed the secuirty hotspot

**Comment by rohityadavcloud** [comment]
Thanks @rp- 
@blueorangutan package

**Comment by DaanHoogland** [comment]
@rp- can you look at the conflicts?

**Comment by rp-** [comment]
@DaanHoogland done and weird, because I just had todo a local rebase. no conflicts.

**Comment by DaanHoogland** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=apache_cloudstack&pullRequest=6682)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=BUG) [C](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=BUG) [2 Bugs](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6682&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6682&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6682&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=CODE_SMELL) [13 Code Smells](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6682&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6682&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6682&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6682&metric=new_duplicated_lines_density&view=list) [2.8% Duplication](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6682&metric=new_duplicated_lines_density&view=list)

@rp- looks mostly good but can you assess these ^^ I think some of them are easy to resolve. Of the bugs ignore the first one but please fix the second (potential NPE)

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rp-** [comment]
@DaanHoogland done, even though the NPE was a false positive (but also my IDEA complained about it)

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by rp-** [comment]
> Packaging result: heavy_check_mark el7 heavy_check_mark el8 heavy_multiplication_x debian heavy_check_mark suse15. SL-JID 4311

@rohityadavcloud is that debian packaging error my fault? where could I look up the error?

**Comment by rohityadavcloud** [comment]
I think it's some intermittent pkging error @rp- I'll rekick them again.
@blueorangutan package

File: debian/rules [hunk 1]
```diff
@@ -42,6 +42,7 @@ override_dh_auto_install:
 	install -D plugins/hypervisors/kvm/target/cloud-plugin-hypervisor-kvm-$(VERSION).jar $(DESTDIR)/usr/share/$(PACKAGE)-agent/lib/
 	install -D plugins/hypervisors/kvm/target/dependencies/* $(DESTDIR)/usr/share/$(PACKAGE)-agent/lib/
 	install -D plugins/storage/volume/storpool/target/cloud-plugin-storage-volume-storpool-$(VERSION).jar $(DESTDIR)/usr/share/$(PACKAGE)-agent/lib/
+	install -D plugins/storage/volume/storpool/target/cloud-plugin-storage-volume-linstor-$(VERSION).jar $(DESTDIR)/usr/share/$(PACKAGE)-agent/lib/
```
Comment (DaanHoogland) [inline debian/rules:45]: @rp- , during the packaging phase the following is reported:
```
install -D plugins/storage/volume/storpool/target/cloud-plugin-storage-volume-linstor-4.18.0.0-shapeblue5122.jar "debian/tmp"/usr/share/cloudstack-agent/lib/
install: cannot stat 'plugins/storage/volume/storpool/target/cloud-plugin-storage-volume-linstor-4.18.0.0-shapeblue5122.jar': No such file or directory
debian/rules:26: recipe for target 'override_dh_auto_install' failed
make[1]: *** [override_dh_auto_install] Error 1
make[1]: Leaving directory '/jenkins/workspace/acs-deb-pkg-builder'
debian/rules:9: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: debian/rules binary gave error exit status 2
Build step 'Execute shell' marked build as failure
```
I think you made a typo here:
```suggestion
	install -D plugins/storage/volume/linstor/target/cloud-plugin-storage-volume-linstor-$(VERSION).jar $(DESTDIR)/usr/share/$(PACKAGE)-agent/lib/
```
???

Comment (rp-) [reply debian/rules]: yep correct, thanks for catching

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]
LGTM - we can't test for linstor related regressions if any and for that @rp- is assumed to have done the due diligence as the vendor/maintainer. The regression tests have passed so we can merge this.

**Pull request merged**